### PR TITLE
Cantera extension to Plasma Flows

### PR DESCRIPTION
**Pull request by AndreaAlberti88**
Cantera has been extended to plasma flows (for details please refer to README_plasma.txt).

This material is based upon work supported by the Department of Energy, National Nuclear Security Administration, under Award Number(s) DE-NA0002374.

XPACC (THE CENTER FOR EXASCALE SIMULATION OF PLASMA-COUPLED COMBUSTION): a thermochemistry library for plasma coupled combustion was needed for simulations. It has been decided to extend Cantera to plasma flows.
1) Interaction potentials updated for neutral-neutral interactions
2) Ionized species introduced 
3) Chapman-Enskog based transport implemented

FUTURE WORK: Cantera needs to be extended to Multi-Temperature (we are now working on this at the University of Illinois at Urbana-Champaign)

### CODE REVIEW

File: include/cantera/Cantera.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+// Include string utility routines
+#include "base/stringUtils.h"
+
+// Include the array object
+#include "base/Array.h"
+
+#endif
+
+
```
Comment (speth) [inline include/cantera/Cantera.h]: "Cantera.h" was deprecated a while back and removed. Files which need any of the referenced includes should just name them directly.

File: include/cantera/base/Array.h [hunk 2]
```diff
@@ -67,6 +67,10 @@ class Array2D
         :  m_data(0), m_nrows(m), m_ncols(n) {
         m_data.resize(n*m);
         std::fill(m_data.begin(), m_data.end(), v);
+
+
+//std::cout << "array costructor 1" << m_nrows << "   " << m_ncols  << std::endl;
+
```
Comment (speth) [inline include/cantera/base/Array.h]: Please avoid adding commented-out code, especially to otherwise-unmodified files.

File: include/cantera/numerics/NonlinearSolver.h [hunk 3]
```diff
@@ -6,8 +6,8 @@
  */
 
 /*
- *  $Date$
- *  $Revision$
+ *  $Date: 2014-01-03 14:57:40 -0600 (Fri, 03 Jan 2014) $
+ *  $Revision: 2644 $
```
Comment (speth) [inline include/cantera/numerics/NonlinearSolver.h]: This change should be removed since it will just make it more difficult to bring the branch up to the present state.

File: include/cantera/thermo/Phase.h [hunk 4]
```diff
@@ -776,6 +789,12 @@ class Phase
     //! Vector of the species names
     std::vector<std::string> m_speciesNames;
 
+/*
+	std::vector<std::string> m_speciesNamesNeutrals;
+        std::vector<std::string> m_speciesNamesPositiveIons;
+        std::vector<std::string> m_speciesNamesNegativeIons;
+*/
+
```
Comment (speth) [inline include/cantera/thermo/Phase.h]: Again, please avoid committing commented out code.

File: include/cantera/thermo/Phase.h [hunk 5]
```diff
@@ -266,6 +266,19 @@ class Phase
     //! Return a const reference to the vector of species names
     const std::vector<std::string>& speciesNames() const;
 
+//    const std::vector<std::string>& speciesNamesNeutrals() const;
+
+    void speciesNamesPositiveIons(std::vector<std::string> & vec, std::vector<int> &vec2);
```
Comment (speth) [inline include/cantera/thermo/Phase.h]: These methods need to be documented (in particular, the meaning of the arguments. I would suggest more descriptive names than "vec" and "vec2").

File: include/cantera/thermo/Phase.h [hunk 6]
```diff
@@ -266,6 +266,19 @@ class Phase
...
+//    const std::vector<std::string>& speciesNamesNegativeIons() const;
+    void speciesNamesNegativeIons(std::vector<std::string> & vec, std::vector<int> &vec2);
+    void speciesNamesNeutrals(std::vector<std::string> & vec, std::vector<int> &vec2);
+
+
+
+
+
+
+
```
Comment (speth) [inline include/cantera/thermo/Phase.h]: Please avoid introducing unnecessary whitespace

File: include/cantera/thermo/Phase.h [hunk 7]
```diff
@@ -785,6 +804,26 @@ class Phase
...
+        std::vector<std::string> spNeut;
+        std::vector<int> spNeutIndex;
+
+
+        std::vector<int> indexNeutNeut;
+        std::vector<int> indexNeutPos;
+        std::vector<int> indexNeutNeg;
+        std::vector<int> indexPosPos;
+        std::vector<int> indexPosNeg;
+        std::vector<int> indexNegNeg;
```
Comment (speth) [inline include/cantera/thermo/Phase.h]: For most classes in Cantera, the convention is that member variables are prefixed with "m_", e.g. "m_indexPosPos". There need to be comments describing what these variables are.

Most member variables in class Phase are "protected" not "public". I don't immediately see a reason that these need to be public.

I'm not sure why member variables for the names of the species of each type are needed. If you have the list of species indices, you can always retrieve the name efficiently.

File: include/cantera/transport/GasTransport.h [hunk 8]
```diff
@@ -104,13 +107,18 @@ class GasTransport : public Transport
      */
     virtual void getMixDiffCoeffsMass(doublereal* const d);
 
+
 protected:
     GasTransport(ThermoPhase* thermo=0);
 
     virtual bool initGas(GasTransportParams& tr);
+
+protected:
```
Comment (speth) [inline include/cantera/transport/GasTransport.h]: Please avoid introducing redundant labels.

File: src/thermo/#ThermoFactory.cpp# [hunk 9]
```diff
@@ -0,0 +1,630 @@
+/**
```
Comment (speth) [inline src/thermo/#ThermoFactory.cpp#]: This file should be removed from the commit.

File: test_problems/test_case_neutralMixture_ChemicalEquilibrium/.depends [hunk 10]
```diff
@@ -0,0 +1,57 @@
+mixture.o: mixture.cpp \
```
Comment (speth) [inline test_problems/test_case_neutralMixture_ChemicalEquilibrium/.depends:1]: This file should not be included in the commit.

File: README_plasma.txt [hunk 11]
```diff
@@ -0,0 +1,33 @@
+University of Illinois at Urbana-Champaign - September 2015
+
+In this branch, Cantera has been extended to Plasma flows.
+
+New capabilities:
```
Comment (speth) [inline README_plasma.txt:5]: Documentation should be added as Doxygen docstrings in the header files and in the Sphinx documentation, both of which are used to produce the online documentation.

**Comment by speth** [comment]
The binary file `test_problems/test_case_neutralMixture_ChemicalEquilibrium/mixture` should be removed from the commit.

I've made a number of specific comments, but these comments apply to many instances of these issues within the code.

It appears that the changes mainly extend the MixTransport class. I think that it might be best to move the additions to MixTransport into a class which is derived from MixTransport, e.g. PlasmaTransport.

**Comment by speth** [comment]
I wanted to note the high likelihood for conflicts between this branch and the work being done by @BangShiuh on #357. Both of these PRs introduce capabilities for dealing with transport of charged species, but (I assume) using different models.

For this reason, I would like to reiterate the suggestion that both of these sets of changes would probably be best implemented by introducing new derived classes of Transport (derived from `GasTransport`, `MixTransport`, or `MultiTransport` as appropriate).

**Comment by BangShiuh** [comment]
The model Andrea used is more accurate. I saw high order of collision integral been used. The ambipolar diffusion method can use the binary diffusion coefficient generated by the model. The main correction to the diffusion flux I added is at StFlow.cpp. So if Andrea get the module finished, I will probably use the transport of his module.

**Comment by speth** [comment]
I'm going to close this PR, since it seems to be stale. A new PR with a cleaner set of changes to the code, based on the current master branch, would be welcome at some point in the future if there's still interest in integrating these capabilities into Cantera.
